File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.java
class Solution {

    public int dfs(int prev, int index, int[] numbers, int target) {
        // condition
        if (index >= numbers.length){
            if (target == prev){
                return 1;
            }
        return 0;
        }
        
        int cur1 = prev + numbers[index];
        int cur2 = prev - numbers[index];
            
        int ans = 0;
        ans += dfs(cur1, index+1, numbers, target);
        ans += dfs(cur2, index+1, numbers, target);
        
        return ans;
    }

    public int solution(int[] numbers, int target) {
        int answer = 0;
        int current = numbers[0];
        
        answer += dfs(current, 1, numbers, target);
        answer += dfs(-current, 1, numbers, target);
        
        return answer;
    }
}

File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.py
# Python Sol with bfs
from collections import deque

def solution(numbers, target):
    answer = 0
    queue = deque()
    
    queue.append((numbers[0], 0))
    queue.append((-numbers[0], 0))
    
    while queue :
        p = queue.popleft()
        if p[1] == len(numbers) -1 :
            if p[0] == target :
                answer += 1
            continue
            
        queue.append((p[0] + numbers[p[1]+1], p[1]+1))
        queue.append((p[0] - numbers[p[1]+1], p[1]+1))
    
    return answer